sdk/v2/core/events.go

```diff
@@ -0,0 +1,361 @@
+	// specified value in the Source field, thereby eliminating the possibility of
+	// gateways maliciously creating events that spoof events from another
+	// gateway.
+	Source string `json:"source,omitempty"`
+	// Type specifies the exact event that has occurred in the upstream system.
+	// Values are opaque and source-specific.
+	Type string `json:"type,omitempty"`
+	// Labels convey additional event details for the purposes of matching Events
+	// to subscribed projects. For instance, no subscribers to the "GitHub" Source
+	// and the "push" Type are likely to want to hear about push events for ALL
+	// repositories. If the "GitHub" gateway labels events with the name of the
+	// repository from which the event originated (e.g. "repo=github.com/foo/bar")
+	// then subscribers can utilize those same criteria to narrow their
+	// subscription from all push events emitted by the GitHub gateway to just
+	// those having originated from a specific repository.
+	Labels Labels `json:"labels,omitempty"`
+	// ShortTitle is an optional, succinct title for the Event, ideal for use in
+	// lists or in scenarios where UI real estate is constrained.
+	ShortTitle string `json:"shortTitle,omitempty"`
+	// LongTitle is an optional, detailed title for the Event.
+	LongTitle string `json:"longTitle,omitempty"`
+	// Git contains git-specific Event details. These can be used to override
+	// similar details defined at the Project level. This is useful for scenarios
+	// wherein an Event may need to convey an alternative source, branch, etc.
+	Git *EventGitConfig `json:"git,omitempty"`
+	// Payload optionally contains Event details provided by the upstream system
+	// that was the original source of the event. Payloads MUST NOT contain
+	// sensitive information. Since Projects SUBSCRIBE to Events, the potential
+	// exists for any Project to express an interest in any or all Events. This
+	// being the case, sensitive details must never be present in payloads. The
+	// common workaround work this constraint (which is also a sensible practice
+	// to begin with) is that event payloads may contain REFERENCES to sensitive
+	// details that are useful only to properly configured Workers.
+	Payload string `json:"payload,omitempty"`
+	// Worker contains details of the worker that will/is/has handle(d) the Event.
+	Worker *Worker `json:"worker,omitempty"`
+}
+
+// MarshalJSON amends Event instances with type metadata so that clients do not
+// need to be concerned with the tedium of doing so.
+func (e Event) MarshalJSON() ([]byte, error) {
+	type Alias Event
+	return json.Marshal(
+		struct {
+			meta.TypeMeta `json:",inline"`
+			Alias         `json:",inline"`
+		}{
+			TypeMeta: meta.TypeMeta{
+				APIVersion: meta.APIVersion,
+				Kind:       "Event",
+			},
+			Alias: (Alias)(e),
+		},
+	)
+}
+
+// EventGitConfig represents git-specific Event details. These may override
+// similar details set at the Project level.
+type EventGitConfig struct {
```

Comment:
I will take a closer look, but I know a few spots where similarly named types exist because in some contexts they need additional fields that aren't appropriate in some _other_ context. But I'll definitely revisit the question of whether it's appropriate to use composition to DRY those cases up a bit.

Reply:
I think all of this was resolved in one of the other conversations.